tensorflow_similarity/augmenters/barlow.py

```diff
@@ -0,0 +1,148 @@
+import tensorflow as tf
+from tensorflow import keras 
+import tensorflow_addons as tfa
+import numpy as np
+from tensorflow_similarity.augmenters.augmenter import Augmenter 
+from typing import Callable, List, Optional, Tuple, Any
+import os
+
+class Augmentation(keras.layers.Layer):
+    def __init__(self):
+        super(Augmentation, self).__init__()
+
+    @tf.function
+    def random_execute(self, prob: float) -> bool:
+        return bool(tf.random.uniform([], minval=0, maxval=1) < prob)
+
+
+class RandomToGrayscale(Augmentation):
+    @tf.function
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+
+        if self.random_execute(0.2):
+            x = tf.image.rgb_to_grayscale(x)
+            x = tf.tile(x, [1, 1, 3])
+        return x
+
+
+class RandomColorJitter(Augmentation):
+    @tf.function
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+        if self.random_execute(0.8):
+            x = tf.image.random_brightness(x, 0.8)
+            x = tf.image.random_contrast(x, 0.4, 1.6)
+            x = tf.image.random_saturation(x, 0.4, 1.6)
+            x = tf.image.random_hue(x, 0.2)
+        return x
+
+
+class RandomFlip(Augmentation):
+    @tf.function
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+        if self.random_execute(0.5):
+            x = tf.image.random_flip_left_right(x)
+        return x
+
+
+class RandomResizedCrop(Augmentation):
+    def __init__(self):
+        super(Augmentation, self).__init__()
+
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+        x_dim = tf.shape(x)
+        x_width = tf.cast(x_dim[0], tf.float32) 
+        x_height = tf.cast(x_dim[1], tf.float32)
+
+        print(x_width)
+        rand_size_width = tf.random.uniform(
+            shape=[],
+            minval=int(0.75 * x_width),
+            maxval=int(1 * x_width),
+            dtype=tf.int32,
+        )
+
+        rand_size_height = tf.random.uniform(
+            shape=[],
+            minval=int(0.75 * x_height),
+            maxval=int(1 * x_height),
+            dtype=tf.int32,
+        )
+
+        crop = tf.image.random_crop(x, (x_width, x_height, 3))
+        crop_resize = tf.image.resize(crop, (x_width, x_height))
+        return crop_resize
+
+
+class RandomSolarize(Augmentation):
+    @tf.function
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+        if self.random_execute(0.2):
+            # flips abnormally low pixels to abnormally high pixels
+            x = tf.where(x < 10, x, 255 - x)
+        return x
+
+
+class RandomBlur(Augmentation):
+    @tf.function
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+        if self.random_execute(0.2):
+            s = np.random.random()
+            return tfa.image.gaussian_filter2d(image=x, sigma=s)
```

Comment:
Makes sense to me. By the way, as tf-similarity implements more contrastive learning algorithms a lot of the augmentation utilities will be shared. I think it would be good to have a single file for augmentation utilities and selectively import parts of the file depending on the needs of the algorithm.

Reply:
Agreed. We added the image augmentations as we needed them for the various algorithms, but it would be good to refactor this into a single module with just the pieces we need. Then we can make the paper specific augmenters using the functions in the module.

Reply:
I'm currently working on this, and have made commit to my fork. Currently testing it.